### DISCUSSION TITLE
Optimizing trajectory aggregator's nearest point detection

### DISCUSSION THREAD

**Discussion by boiko** [anitagraser/movingpandas#202] [Ideas]
By running movingpandas under cProfile, it is possible to see that one of the biggest time consuming task of the trajectory aggregator is to find the nearest cells of trajectory points:
![image](https://user-images.githubusercontent.com/254625/145135596-039cd20f-8f06-487c-8637-0398d377e1a6.png)

By inserting the cells in a more efficient data structure (skearn's [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) using kd-tree for instance), a lot less time is spent on the same task (this is a rudimentary prototype to serve as a proof of concept):
![image](https://user-images.githubusercontent.com/254625/145135396-27a5fbcf-d273-4364-8d6c-34187a420cc2.png)

This is not the only option for implementing this, here are some other suggestions:
* [scipy.spatial.KDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.html)
* [geopandas' SpatialIndex.nearest](https://geopandas.org/en/stable/docs/reference/api/geopandas.sindex.SpatialIndex.nearest.html). There is one concern regarding this one for now though: it only works efficiently when using `pygeos` which, just like `shapely`, uses the GEOS C library, and when the shapely and pygeos are built using different versions of this library, it becomes inneficient. This will probably be solved soon since pygeos got merged into shapely, but it seems we still need versions lower than 1.8 because of geoviews.

The purpose of this issue is to collect feedback and discuss these and maybe other possible approaches for the task before implementing it.

**Comment by anitagraser** [comment]
I'm tempted to wait for the new geoviews version (currently 1.9.2). The ticket for updated shapely support is already closed https://github.com/holoviz/geoviews/issues/533

**Reply by boiko** [reply]
Very late reply... I guess it makes sense. If the performance difference is not too big, the simplicity of the code and not having to add extra dependencies is worth it.

I will keep an eye on it and do some measurements when geopandas get adjusted to work with latest shapely (which includes pygeos).
